src/solver/cumulative_reduction/forward_cumulative_reduction_contiguous_lastdim.cpp

```diff
@@ -0,0 +1,146 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/datatype.hpp>
+#include <miopen/kernel_build_params.hpp>
+#include <miopen/cumulative_reduction/invoke_params.hpp>
+#include <miopen/cumulative_reduction/solvers.hpp>
+#include <miopen/cumulative_reduction/utils.hpp>
+
+#define FLOAT_ACCUM float
+
+#define LOCAL_SIZE_MAX 256
+#define LOCAL_SIZE_MIN 64
+
+namespace miopen {
+
+namespace solver {
+
+namespace cumulative_reduction {
+
+bool IsImprovementOverROCm(const ExecutionContext& /*context*/,
+                           const miopen::cumulative_reduction::ForwardProblemDescription& problem)
+{
+    if(problem.GetInputDesc().GetLengths()[problem.GetDim()] > LOCAL_SIZE_MAX)
+        return false;
+    return true;
+}
+
+bool ForwardContiguousLastDim::IsApplicable(
+    const ExecutionContext& context,
+    const miopen::cumulative_reduction::ForwardProblemDescription& problem) const
+{
+    if(!IsImprovementOverROCm(context, problem))
+        return false;
+    if(!problem.IsAllPacked())
+        return false;
+    if(problem.GetInputDesc().GetStrides()[problem.GetDim()] != 1)
+        return false;
+    if(problem.GetInputDesc().GetSize() == problem.GetOutputDesc().GetSize() &&
+       problem.GetOutputDesc().GetStrides()[problem.GetDim()] != 1)
+        return false;
+    if(problem.GetInputDesc().GetSize() == problem.GetIndicesDesc().GetSize() &&
+       problem.GetIndicesDesc().GetStrides()[problem.GetDim()] != 1)
+        return false;
+    return true;
+}
+
+ConvSolution ForwardContiguousLastDim::GetSolution(
+    const ExecutionContext& /*context*/,
+    const miopen::cumulative_reduction::ForwardProblemDescription& problem) const
+{
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    auto dtype        = problem.GetInputDesc().GetType();
+    auto input_dtype  = miopen::GetDataType(problem.GetInputDesc().GetType());
+    auto output_dtype = miopen::GetDataType(problem.GetOutputDesc().GetType());
+    auto cum_op       = problem.GetCumOp();
+
+    auto size       = problem.GetInputDesc().GetElementSize();
+    auto inner_size = problem.GetInputDesc().GetLengths()[problem.GetDim()];
+    auto outer_size = size / inner_size;
+
+    auto local_size = LOCAL_SIZE_MIN;
+    while(local_size < inner_size)
+        local_size *= 2;
+
+    auto build_params = KernelBuildParameters{
+        {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+        {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+        {"MIOPEN_USE_FP64", static_cast<int>(dtype == miopenDouble)},
+        {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+        {"INPUT_TYPE", input_dtype == "bfloat16" ? "ushort" : input_dtype},
```

Comment:
Here you are only setting `dtype` based on input types, not output. This ties in to the other comment in the kernel (do you ever expect mixed input/output types). Because if you are, this is wrong, and if you are not, this is unnecessary